lib/livebook/zta/teleport.ex

```diff
@@ -0,0 +1,90 @@
+defmodule Livebook.ZTA.Teleport do
+  use GenServer
+  require Logger
+
+  defstruct [:name, :req_options, :jwks]
+  @fields %{"sub" => :id, "username" => :username}
+
+  @assertion "teleport-jwt-assertion"
+  def start_link(opts) do
+    url = opts[:identity][:key]
+    options = [req_options: [url: url]]
+    GenServer.start_link(__MODULE__, options, name: opts[:name])
+  end
+
+  @impl true
+  def init(options) do
+    state = struct!(__MODULE__, options)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_call(:get_jwks, _, state) do
+    try do
+      jwks =
+        Req.request!(state.req_options).body["keys"]
+        |> JOSE.JWK.from_map()
+
+      {:reply, jwks, %{state | jwks: jwks}}
+    rescue
+      error ->
+        Logger.debug(
+          "[#{inspect(__MODULE__)}] requested #{inspect(state.req_options)} and got error #{inspect(error)}"
+        )
+
+        {:reply, [], state}
+    end
+  end
+
+  def authenticate(name, conn, _) do
+    token = Plug.Conn.get_req_header(conn, @assertion)
+
+    jwks = GenServer.call(name, :get_jwks, :infinity)
+
+    {conn, authenticate_user(token, jwks)}
+  end
+
+  defp authenticate_user(token, jwks) do
+    with [encoded_token] <- token,
+         {:ok, token} <- verify_token(encoded_token, jwks),
```

Comment:
```suggestion
         {:ok, %{fields: %{"iat" => iat, "exp" => exp, "nbf" => nbf}}} <- verify_token(encoded_token, jwks),
```

Then you can pass these values directly to `verify_timestamps` and remove one clause from verify_timestamps.